### PR TITLE
Add HyperGeometric Distribution to pymc3.distributions.discrete #3504

### PR DESCRIPTION
**Pull request by Harivallabha**
- Support HyperGeometric Distribution as part of distributions/discrete
- Added tests
- Added docstrings

@twiecki @ricardov94. Hope this helps. Please feel free to modify/add on top of this, if this is useful. If not, please ignore 😬

### CODE REVIEW

**Comment by Harivallabha** [comment]
I can see that `scipy.stats.hypergeom` and the above `pymc3.distributions.discrete.HyperGeometric` give the same result for usual values. 

However, when I test for something like: M = 3, n = 3, N = 4, and k = 1
` scipy.stats.hypergeom.logpmf(1, 3, 3, 4)` gives nan

And, 
`with pm.Model() as model:`
    `z = pm.HyperGeometric('z', 3, 3, 4)`
`model.logp({'z':1})` gives -inf

**Comment by fonnesbeck** [comment]
Thanks for the PR! Can you please run the Black formatter on this?

**Comment by Harivallabha** [comment]
@fonnesbeck Sure! I should tell you that running the Black formatter on discrete.py, test_distributions.py, and test_distributions_random.py formats some code that I haven't touched as well :sweat_smile:

**Comment by fonnesbeck** [comment]
No problem. It likely slipped through on a previous PR.

**Comment by Harivallabha** [comment]
Okay, done! The Black formatter has reached in and changed stuff in a lottt of places (Exa: change all ' '  to " "). So lemme know if I should revert this because this is probably not gonna be easy for you to review :sweat_smile:

File: pymc3/distributions/discrete.py [hunk 1]
```diff
@@ -819,6 +820,110 @@ def _repr_latex_(self, name=None, dist=None):
...
+            values are desired the values must be provided in a numpy array or theano tensor
+        Returns
+        -------
+        TensorVariable
+        """
+        N = self.N
+        k = self.k
+        n = self.n
+        return bound(binomln(k, value) + binomln(N - k, n - value) - binomln(N, n),
+                     0 <= k, k <= N, 0 <= n, 0 <= N, n - N + k <= value, 0 <= value,
```
Comment (fonnesbeck) [inline pymc3/distributions/discrete.py]: You are already testing that 0 <= k and k <= N, so you should not have to include 0 <= N

Comment (Harivallabha) [reply pymc3/distributions/discrete.py]: Ah, point! Removing the unnecessary condition.

**Comment by fonnesbeck** [comment]
As for the difference between your implementation and SciPy's, theirs uses a different formulat for the logprob:

```python
    def _logpmf(self, k, M, n, N):
        tot, good = M, n
        bad = tot - good
        result = (betaln(good+1, 1) + betaln(bad+1, 1) + betaln(tot-N+1, N+1) -
                  betaln(k+1, good-k+1) - betaln(N-k+1, bad-N+k+1) -
                  betaln(tot+1, 1))
        return result
```

So, you can either change yours to match this, or address the test another way.

**Comment by MarcoGorelli** [comment]
> Okay, done! The Black formatter has reached in and changed stuff in a lottt of places (Exa: change all ' ' to " "). So lemme know if I should revert this because this is probably not gonna be easy for you to review sweat_smile

FWIW, I think this is a fair point...would PyMC3 take a large PR that does nothing except for applying `black` everywhere and also adds it to CI / pre-commit? Because otherwise reviews do become hard. Conversely, once `black` has been applied everywhere, diffs do tend to become smaller.

EDIT
----
taken forward in  #4109

**Comment by Harivallabha** [comment]
@fonnesbeck Thanks! Modified the implementation to match the scipy one.

**Comment by fonnesbeck** [comment]
Looks like you need a rebase to address the conflicts.

File: pymc3/distributions/discrete.py [hunk 2]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+            plt.plot(x, pmf, '-o', label='n = {}'.format(n))
+        plt.plot(x, pmf, '-o', label='N = {}'.format(N))
+        plt.plot(x, pmf, '-o', label='k = {}'.format(k))
+        plt.xlabel('x', fontsize=12)
+        plt.ylabel('f(x)', fontsize=12)
+        plt.legend(loc=1)
+        plt.show()
+
+    ========  =============================
+    Support   :math:`x \in \mathbb{N}_{>0}`
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:889]: Are you sure this is right? On the [Wikipedia page](https://en.wikipedia.org/wiki/Hypergeometric_distribution), support is given as ``x in [max(0, n - N + k), min(k, n)]``.

File: pymc3/distributions/discrete.py [hunk 3]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
-                                                get_variable_name(p))
+        name = r"\text{%s}" % name
+        return r"${} \sim \text{{Geometric}}(\mathit{{p}}={})$".format(
+            name, get_variable_name(p)
+        )
+
+
+class HyperGeometric(Discrete):
+    R"""
+    Hypergeometric log-likelihood.
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:861]:
```suggestion
    Discrete hypergeometric distribution.
```

Isn't this better?

File: pymc3/distributions/discrete.py [hunk 4]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+
+    The probability of x successes in a sequence of n Bernoulli
+    trials (That is, sample size = n) - where the population
+    size is N, containing a total of k successful individuals.
+    The process is carried out without replacement.
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:866]:
```suggestion
    The probability of :math:`x` successes in a sequence of :math:`n` bernoulli
    trials taken without replacement from a population of :math:`N` objects,
    containing :math:`k` good (or successful or Type I) objects.
```

Nitpick. Not a blocking comment.

File: pymc3/distributions/discrete.py [hunk 5]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+        plt.plot(x, pmf, '-o', label='k = {}'.format(k))
+        plt.xlabel('x', fontsize=12)
+        plt.ylabel('f(x)', fontsize=12)
+        plt.legend(loc=1)
+        plt.show()
+
+    ========  =============================
+    Support   :math:`x \in \mathbb{N}_{>0}`
+    Mean      :math:`\dfrac{n.k}{N}`
+    Variance  :math:`\dfrac{(N-n).n.k.(N-k)}{(N-1).N^2}`
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:891]:
```suggestion
    Mean      :math:`\dfrac{nk}{N}`
    Variance  :math:`\dfrac{(N-n)nk(N-k)}{(N-1)N^2}`
```

File: pymc3/distributions/discrete.py [hunk 6]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+        Number of samples drawn from the population
+    k : integer
+        Number of successful individuals in the population
+    """
+
+    def __init__(self, N, k, n, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.N = N = tt.as_tensor_variable(intX(N))
+        self.k = k = tt.as_tensor_variable(intX(k))
+        self.n = n = tt.as_tensor_variable(intX(n))
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:908]:
```suggestion
        self.N = intX(N)
        self.k = intX(k)
        self.n = intX(n)
```
Nitpick. Not a blocking comment.

File: pymc3/distributions/discrete.py [hunk 7]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+        self.mode = intX(tt.floor((n + 1) * (k + 1) / (N + 2)))
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:909]: If the inputs were not integers, the current formula may produce wrong results.
Example: If `N = 10.6`, `n = 2.9`, `k = 6.1`, current formula will produce result `2` but the correct result is `1`.
Can you please change it as follows?

```suggestion
        self.mode = intX(tt.floor((self.n + 1) * (self.k + 1) / (self.N + 2)))
```

File: pymc3/distributions/discrete.py [hunk 8]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+            Desired size of random sample (returns one sample if not
+            specified).
+        Returns
+        -------
+        array
+        """
+        N, n, k = draw_values([self.N, self.n, self.k], point=point, size=size)
+        return generate_samples(
+            np.random.hypergeometric, N, n, k, dist_shape=self.shape, size=size
+        )
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:929]: are you sure this is right? from the [numpy's documentation page of hypergeometric distribution](https://numpy.org/doc/stable/reference/random/generated/numpy.random.hypergeometric.html), `np.random.hypergeometric` accepts arguments `ngood, nbad, nsample` while you are passing in `ntotal (N), nsample (n), ngood (k)`. You should modify the call as:

```suggestion
        return generate_samples(
            np.random.hypergeometric, k, N - k, n, dist_shape=self.shape, size=size
        )
```

File: pymc3/distributions/discrete.py [hunk 9]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+        self.N = N = tt.as_tensor_variable(intX(N))
+        self.k = k = tt.as_tensor_variable(intX(k))
+        self.n = n = tt.as_tensor_variable(intX(n))
+        self.mode = intX(tt.floor((n + 1) * (k + 1) / (N + 2)))
+
+    def random(self, point=None, size=None):
+        r"""
+        Draw random values from HyperGeometric distribution.
+        Parameters
+        ----------
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:915]:
```suggestion
        Draw random values from HyperGeometric distribution.

        Parameters
        ----------
```

Nitpick. Not a blocking comment

File: pymc3/distributions/discrete.py [hunk 10]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+    Hypergeometric log-likelihood.
+
+    The probability of x successes in a sequence of n Bernoulli
+    trials (That is, sample size = n) - where the population
+    size is N, containing a total of k successful individuals.
+    The process is carried out without replacement.
+
+    The pmf of this distribution is
+    .. math:: f(x \mid N, n, k) = \frac{\binom{k}{x}\binom{N-k}{n-x}}{\binom{N}{n}}
+    .. plot::
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:870]:
```suggestion
    The pmf of this distribution is

    .. math:: f(x \mid N, n, k) = \frac{\binom{k}{x}\binom{N-k}{n-x}}{\binom{N}{n}}

    .. plot::
```
Nitpick. Not a blocking comment

File: pymc3/distributions/discrete.py [hunk 11]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+        ----------
+        point : dict, optional
+            Dict of variable values on which random values are to be
+            conditioned (uses default point if not specified).
+        size : int, optional
+            Desired size of random sample (returns one sample if not
+            specified).
+        Returns
+        -------
+        array
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:924]:
```suggestion
            specified).

        Returns
        -------
        array
```
Nitpick. Not a blocking comment

File: pymc3/distributions/discrete.py [hunk 12]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+        N, n, k = draw_values([self.N, self.n, self.k], point=point, size=size)
+        return generate_samples(
+            np.random.hypergeometric, N, n, k, dist_shape=self.shape, size=size
+        )
+
+    def logp(self, value):
+        r"""
+        Calculate log-probability of HyperGeometric distribution at specified value.
+        Parameters
+        ----------
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:935]:
```suggestion
        Calculate log-probability of HyperGeometric distribution at specified value.

        Parameters
        ----------
```
Nitpick. Not a blocking comment

File: pymc3/distributions/discrete.py [hunk 13]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+        value : numeric
+            Value(s) for which log-probability is calculated. If the log probabilities for multiple
+            values are desired the values must be provided in a numpy array or theano tensor
+        Returns
+        -------
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:940]:
```suggestion
            values are desired the values must be provided in a numpy array or theano tensor

        Returns
        -------
```
Nitpick. Not a blocking comment

File: pymc3/distributions/discrete.py [hunk 14]
```diff
@@ -807,16 +844,127 @@ def logp(self, value):
...
+        bad = tot - good
+        result = (
+            betaln(good + 1, 1)
+            + betaln(bad + 1, 1)
+            + betaln(tot - n + 1, n + 1)
+            - betaln(value + 1, good - value + 1)
+            - betaln(n - value + 1, bad - n + value + 1)
+            - betaln(tot + 1, 1)
+        )
+        return result
```
Comment (tirthasheshpatel) [inline pymc3/distributions/discrete.py:956]: You should mask the invalid entries with `-inf` before returning the result. I have used the inequality of support from the Wikipedia page. If that comment is wrong, please change the `lower` and `upper` bound conditions according to the formula of support you use.

```suggestion
        # value in [max(0, n - N + k), min(k, n)]
        lower = tt.switch(tt.gt(n - N + k, 0), n - N + k, 0)
        upper = tt.switch(tt.lt(k, n), k, n)
        nonint_value = (value != intX(tt.floor(value)))
        return bound(result, lower <= value, value <= upper, nonint_value)
```

There is also an additional condition checking whether all the values are integers. scipy checks it but haven't seen pymc3 use it anywhere. It should ideally be there but has multiple numerical problems (like `1e-17` is considered non integer value). Please avoid if you think the condition is better avoided.

**Review by tirthasheshpatel — COMMENTED** [review]
Thanks, @Harivallabha for implementing this distribution. I have been working on adding `NegativeHypergeometric` and `MultivariateHypergeometric` distributions to pymc3 so I thought I would help with this too. Hope these comments are helpful :)

Please tell me if I have missed anything.

File: pymc3/distributions/discrete.py (hunk 7 above)
Comment (tirthasheshpatel) [reply pymc3/distributions/discrete.py:909]: Seen other distributions and it seems like a lot of them have this problem... You can ignore this for consistency with other distributions.

File: pymc3/distributions/discrete.py (hunk 8 above)
Comment (tirthasheshpatel) [reply pymc3/distributions/discrete.py:929]: or you can just use scipy's hypergeometric distribution

```python
    return generate_samples(
            st.hypergeom.rvs, N, k, n, dist_shape=self.shape, size=size
        )
```

**Comment by Spaak** [comment]
@Harivallabha thanks for the PR, we're trying to get this in release 3.10. For that, could you please:

- [ ] Go over @tirthasheshpatel's comments and mark resolved if so
- [ ] Rebase onto master (hopefully not too painful...) so the tests can run and we can merge
- [ ] Simplify (or probably simply remove) the `_repr_latex_` implementations (probably where many of the conflicts are from) (see #4065 and #4076 for a change to how LaTeX and str representations are implemented)

**Comment by Spaak** [comment]
Thanks, let's continue in #4108 then.

**Pull request closed without merging**